### DISCUSSION TITLE
Make the "out" modifier when calling a method optional if the result is being discarded

### DISCUSSION THREAD

**Discussion by ekolis** [dotnet/csharplang#2778] [Language Ideas] [answered]
In this function:

```csharp
public static bool IsBoolean(string s)
{
	return bool.TryParse(s, out _);
}
```

the `out` modifier is superfluous as the result is being discarded anyway. It would be easier to write as:

```csharp
public static bool IsBoolean(string s)
{
	return bool.TryParse(s, _);
}
```
or maybe even

```csharp
public static bool IsBoolean(string s)
{
	return bool.TryParse(s);
}
```

**Comment by CyrusNajmabadi** [comment, marked as answer]
> ```c#
> public static bool IsBoolean(string s)
> {
> 	return bool.TryParse(s);
> }
> ```

This would make it a breaking change then for `bool` to add a `TryParse` function that didn't have an out param in the future.

**Comment by DavidArno** [comment]
Whilst @CyrusNajmabadi is right that the second example would cause problems, I think there is merit in the other version:
```cs
public static bool IsBoolean(string s)
{
	return bool.TryParse(s, _);
}
```
The need for an `out` with the discard does seem overkill.

**Comment by uygary** [comment]
`out` is part of the signature, and it's quite useful for method resolution and explicitly knowing which overload will be invoked.

```
void Main()
{
	var foo = new Foo();
	var x = 4;
	var _ = 2;
	
	foo.Bar(x, _);
	Console.WriteLine(_);
	
	foo.Bar(x, out _);
	Console.WriteLine(_);
}

public class Foo
{
	public void Bar(int x, out int y)
	{
		y = x;
		Console.WriteLine("Yes.");
	}
	
	public void Bar(int x, int y)
	{
		Console.WriteLine("Nope.");
	}
}
```

Building method resolution based on whether a parameter is used down the line or not... I really don't think this is a good idea to being with.

**Comment by DavidArno** [comment]
@uygary,

You make a good point. Sadly, the team didn't make using `_` as a variable a compilation error when they introduced discards. So dropping the `out` on an out parameter discard could cause confusion. 


So this is a nice idea that likely won't happen.
